### DISCUSSION TITLE
AbstractEpollStreamChannel$EpollStreamUnsafe MemLeak

### DISCUSSION THREAD

**Discussion by Eini-AXA** [netty/netty#11761] [Q&A] [answered]
We got the following error message after running our application for some time.

### Actual behavior

```
LEAK: ByteBuf.release() was not called before it's garbage-collected. See https://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
Created at:
	io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:402)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:178)
	io.netty.channel.unix.PreferredDirectByteBufAllocator.ioBuffer(PreferredDirectByteBufAllocator.java:53)
	io.netty.channel.DefaultMaxMessagesRecvByteBufAllocator$MaxMessageHandle.allocate(DefaultMaxMessagesRecvByteBufAllocator.java:114)
	io.netty.channel.epoll.EpollRecvByteAllocatorHandle.allocate(EpollRecvByteAllocatorHandle.java:75)
	io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:780)
	io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:480)
	io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:378)
	io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
	io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.base/java.lang.Thread.run(Thread.java:834)
```
### Possible Cause

https://github.com/netty/netty/blob/88319d422eef8efd6177b55dc43e67091f8aa035/transport-native-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java#L761-L812

The byte buffer is set to null on line 796, but it is never released. And in each iteration we create a new byte buffer.

### Netty version
4.1.67 Final

### JVM version (e.g. `java -version`)
openjdk version "11.0.5" 2019-10-15 LTS
OpenJDK Runtime Environment 18.9 (build 11.0.5+10-LTS)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.5+10-LTS, mixed mode, sharing)

### OS version (e.g. `uname -a`)
Linux  3.10.0-1160.36.2.el7.x86_64 #1 SMP Thu Jul 8 02:53:40 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

**Comment by chrisvest** [comment]
In the "Possible Cause" snippet, the `pipeline.fireChannelRead(byteBut)` call takes on responsibility of releasing the buffer.

**Comment by NiteshKant** [comment]
Converting to GH discussion

**Comment by pravussum** [comment]
I face the same issue - byteBuf is set to null but still has a reference count > 0 after the call to `fireChannelRead()`, eventually leading to memory leaks being reported. 
Even knowing the `fireChannelRead()` call takes responsibility to release the buffer, I'm at a loss how to continue investigation here... Would be good to have the information of who touched the buffer in which order. 

Netty version is 4.1.68.Final
This happens in a Spring boot Webclient asynchronously streaming the response of a long-running request using backpressure.

**Comment by normanmaurer** [comment, marked as answer]
@pravussum the handler in the `ChannelPipeline` that "consumes" the buffer is responsible to call `release()`

**Reply by metolinamy** [reply]
> @pravussum the handler in the `ChannelPipeline` that "consumes" the buffer is responsible to call `release()`

Help

**Comment by Eini-AXA** [comment]
We will check our InputHandlers if one of them does not release the buffer correctly.

**Reply by metolinamy** [reply]
> We will check our InputHandlers if one of them does not release the buffer correctly.
